File: apps/web/src/lib/scanner.ts
import { numeric } from "./format";
import type { ScannerRow } from "./types";

export interface Filters {
  market: string;
  weapon: string;
  skin: string;
  exterior: string;
  currency: string;
  patternType: string;
  minPrice: string;
  maxPrice: string;
  minProfit: string;
  minRoi: string;
  maxFloat: string;
  paintSeed: string;
  minScore: string;
  minLiquidity: string;
  minConfidence: string;
  maxRisk: string;
  maxSpread: string;
}

export type Sort = "opportunity" | "roi" | "profit" | "price" | "float" | "liquidity" | "risk" | "confidence" | "spread" | "discount" | "recent";
export const emptyFilters: Filters = { market: "", weapon: "", skin: "", exterior: "", currency: "", patternType: "", minPrice: "", maxPrice: "", minProfit: "", minRoi: "", maxFloat: "", paintSeed: "", minScore: "", minLiquidity: "", minConfidence: "", maxRisk: "", maxSpread: "" };

function matchesBound(value: string | number | null, bound: string, minimum: boolean): boolean {
  if (bound.trim() === "") return true;
  const threshold = numeric(bound);
  const number = numeric(value);
  if (threshold === null || number === null) return false;
  return minimum ? number >= threshold : number <= threshold;
}

export function filterAndSort(rows: ScannerRow[], filters: Filters, sort: Sort): ScannerRow[] {
  const filtered = rows.filter((row) =>
    (!filters.market || row.platform === filters.market) &&
    (!filters.weapon || row.weapon === filters.weapon) &&
    (!filters.skin || (row.skin ?? row.market_hash_name).toLocaleLowerCase("fr").includes(filters.skin.toLocaleLowerCase("fr"))) &&
    (!filters.exterior || row.exterior === filters.exterior) &&
    (!filters.currency || row.currency_original === filters.currency) &&
    (!filters.patternType || (filters.patternType === "doppler" ? row.doppler_phase !== null : filters.patternType === "fade" ? row.fade_percentage !== null : row.stickers.length > 0)) &&
    matchesBound(row.price_eur_reference, filters.minPrice, true) &&
    matchesBound(row.price_eur_reference, filters.maxPrice, false) &&
    matchesBound(row.potential_profit_eur, filters.minProfit, true) &&
    matchesBound(row.roi, filters.minRoi, true) &&
    matchesBound(row.float_value, filters.maxFloat, false) &&
    (!filters.paintSeed || row.paint_seed === Number(filters.paintSeed)) &&
    matchesBound(row.opportunity_score, filters.minScore, true) &&
    matchesBound(row.liquidity, filters.minLiquidity, true) &&
    matchesBound(row.confidence, filters.minConfidence, true) &&
    matchesBound(row.risk_score, filters.maxRisk, false) &&
    matchesBound(row.spread_percent, filters.maxSpread, false)
  );
  const fields: Partial<Record<Sort, keyof ScannerRow>> = { opportunity: "opportunity_score", roi: "roi", profit: "potential_profit_eur", price: "price_eur_reference", float: "float_value", liquidity: "liquidity", risk: "risk_score", confidence: "confidence", spread: "spread_percent" };
  const ascending = sort === "price" || sort === "float" || sort === "risk" || sort === "spread";
  return filtered.sort((a, b) => {
    const left = sortValue(a, sort, fields[sort]);
    const right = sortValue(b, sort, fields[sort]);
    if (left === null) return right === null ? a.id.localeCompare(b.id) : 1;
    if (right === null) return -1;
    return (ascending ? left - right : right - left) || a.id.localeCompare(b.id);
  });
}

function sortValue(row: ScannerRow, sort: Sort, field?: keyof ScannerRow): number | null {
  if (sort === "recent") {
    const value = Date.parse(row.observed_at);
    return Number.isFinite(value) ? value : null;
  }
  if (sort === "discount") {
    const price = numeric(row.price_eur_reference);
    const reference = numeric(row.estimated_value_eur);
    return price === null || reference === null ? null : reference - price;
  }
  return field ? numeric(row[field] as string | number | null) : null;
}

File: apps/api/app/services/scanner.py
from dataclasses import dataclass
from decimal import Decimal
from math import ceil
from typing import Any, cast

from sqlalchemy import ColumnElement, Select, and_, asc, desc, func, select
from sqlalchemy.orm import Session, selectinload

from app.core.config import Settings
from app.models import CS2Item, ListingAnalysisSnapshot, MarketListing
from app.schemas.api import Mode, Platform, ScannerFacets, ScannerPage, ScannerSort
from app.services.analysis import row_from_snapshot, row_without_snapshot
from app.services.freshness import fresh_listing


@dataclass(frozen=True)
class ScannerQuery:
    mode: Mode = "live"
    page: int = 1
    page_size: int = 50
    sort: ScannerSort = "opportunity"
    market: Platform | None = None
    weapon: str | None = None
    skin: str | None = None
    exterior: str | None = None
    currency: str | None = None
    pattern_type: str | None = None
    min_price: Decimal | None = None
    max_price: Decimal | None = None
    min_profit: Decimal | None = None
    min_roi: Decimal | None = None
    max_float: Decimal | None = None
    paint_seed: int | None = None
    min_score: int | None = None
    min_liquidity: int | None = None
    min_confidence: int | None = None
    max_risk: int | None = None
    max_spread: Decimal | None = None
    market_hash_name: str | None = None
    paint_seeds: tuple[int, ...] = ()
    doppler_phase: str | None = None


def build_scanner_page(
    session: Session, query: ScannerQuery, settings: Settings | None = None
) -> ScannerPage:
    statement = _filtered_statement(query, settings)
    total = int(
        session.scalar(
            select(func.count()).select_from(
                statement.with_only_columns(MarketListing.id).order_by(None).subquery()
            )
        )
        or 0
    )
    ordered = _apply_order(statement, query.sort)
    records = session.execute(
        ordered.offset((query.page - 1) * query.page_size).limit(query.page_size)
    ).all()
    items = [
        row_from_snapshot(listing, snapshot)
        if snapshot is not None
        else row_without_snapshot(listing)
        for listing, snapshot in records
    ]
    missing = sum(snapshot is None for _, snapshot in records)
    warnings = [f"{missing} analyse(s) attendent la prochaine synchronisation."] if missing else []
    return ScannerPage(
        mode=query.mode,
        items=items,
        total=total,
        page=query.page,
        page_size=query.page_size,
        pages=ceil(total / query.page_size) if total else 0,
        facets=_facets(session, query.mode),
        warnings=warnings,
    )


def _filtered_statement(
    query: ScannerQuery,
    settings: Settings | None = None,
) -> Select[tuple[MarketListing, ListingAnalysisSnapshot | None]]:
    statement = (
        select(MarketListing, ListingAnalysisSnapshot)
        .join(CS2Item, MarketListing.item_id == CS2Item.id)
        .outerjoin(
            ListingAnalysisSnapshot,
            and_(
                ListingAnalysisSnapshot.mode == MarketListing.mode,
                ListingAnalysisSnapshot.listing_id == MarketListing.id,
            ),
        )
        .options(selectinload(MarketListing.item).selectinload(CS2Item.stickers))
        .where(MarketListing.mode == query.mode, MarketListing.status == "ACTIVE")
    )
    if query.market:
        statement = statement.where(MarketListing.platform == query.market)
    if query.mode == "live":
        statement = statement.where(fresh_listing(settings or Settings(_env_file=None)))
    if query.market_hash_name:
        statement = statement.where(CS2Item.market_hash_name == query.market_hash_name)
    if query.paint_seeds:
        statement = statement.where(CS2Item.paint_seed.in_(query.paint_seeds))
    if query.doppler_phase:
        statement = statement.where(
            func.lower(CS2Item.doppler_phase) == query.doppler_phase.lower()
        )
    if query.weapon:
        statement = statement.where(CS2Item.weapon == query.weapon)
    if query.skin:
        statement = statement.where(CS2Item.skin.icontains(query.skin, autoescape=True))
    if query.exterior:
        statement = statement.where(CS2Item.exterior == query.exterior)
    if query.currency:
        statement = statement.where(MarketListing.currency_original == query.currency)
    if query.pattern_type == "doppler":
        statement = statement.where(CS2Item.doppler_phase.is_not(None))
    elif query.pattern_type == "fade":
        statement = statement.where(CS2Item.fade_percentage.is_not(None))
    elif query.pattern_type == "sticker":
        statement = statement.where(CS2Item.stickers.any())
    if query.min_price is not None:
        statement = statement.where(MarketListing.price_eur_reference >= query.min_price)
    if query.max_price is not None:
        statement = statement.where(MarketListing.price_eur_reference <= query.max_price)
    if query.min_profit is not None:
        statement = statement.where(
            ListingAnalysisSnapshot.potential_profit_eur >= query.min_profit
        )
    if query.min_roi is not None:
        statement = statement.where(ListingAnalysisSnapshot.roi >= query.min_roi)
    if query.max_float is not None:
        statement = statement.where(CS2Item.float_value <= query.max_float)
    if query.paint_seed is not None:
        statement = statement.where(CS2Item.paint_seed == query.paint_seed)
    if query.min_score is not None:
        statement = statement.where(ListingAnalysisSnapshot.opportunity_score >= query.min_score)
    if query.min_liquidity is not None:
        statement = statement.where(ListingAnalysisSnapshot.liquidity_score >= query.min_liquidity)
    if query.min_confidence is not None:
        statement = statement.where(ListingAnalysisSnapshot.confidence >= query.min_confidence)
    if query.max_risk is not None:
        statement = statement.where(ListingAnalysisSnapshot.risk_score <= query.max_risk)
    if query.max_spread is not None:
        statement = statement.where(ListingAnalysisSnapshot.spread_percent <= query.max_spread)
    return cast(Select[tuple[MarketListing, ListingAnalysisSnapshot | None]], statement)


def _apply_order(
    statement: Select[tuple[MarketListing, ListingAnalysisSnapshot | None]],
    sort: ScannerSort,
) -> Select[tuple[MarketListing, ListingAnalysisSnapshot | None]]:
    fields: dict[ScannerSort, ColumnElement[Any]] = {
        "opportunity": desc(ListingAnalysisSnapshot.opportunity_score).nulls_last(),
        "roi": desc(ListingAnalysisSnapshot.roi).nulls_last(),
        "profit": desc(ListingAnalysisSnapshot.potential_profit_eur).nulls_last(),
        "price": asc(MarketListing.price_eur_reference).nulls_last(),
        "float": asc(CS2Item.float_value).nulls_last(),
        "liquidity": desc(ListingAnalysisSnapshot.liquidity_score).nulls_last(),
        "risk": asc(ListingAnalysisSnapshot.risk_score).nulls_last(),
        "confidence": desc(ListingAnalysisSnapshot.confidence).nulls_last(),
        "spread": asc(ListingAnalysisSnapshot.spread_percent).nulls_last(),
        "discount": desc(
            ListingAnalysisSnapshot.estimated_value_eur - MarketListing.price_eur_reference
        ).nulls_last(),
        "recent": desc(MarketListing.observed_at),
    }
    return statement.order_by(fields[sort], MarketListing.observed_at.desc(), MarketListing.id)


def _facets(session: Session, mode: Mode) -> ScannerFacets:
    base = (MarketListing.mode == mode, MarketListing.status == "ACTIVE")
    markets = session.scalars(
        select(MarketListing.platform).where(*base).distinct().order_by(MarketListing.platform)
    )
    weapons = session.scalars(
        select(CS2Item.weapon)
        .join(MarketListing, MarketListing.item_id == CS2Item.id)
        .where(*base, CS2Item.weapon.is_not(None))
        .distinct()
        .order_by(CS2Item.weapon)
    )
    exteriors = session.scalars(
        select(CS2Item.exterior)
        .join(MarketListing, MarketListing.item_id == CS2Item.id)
        .where(*base, CS2Item.exterior.is_not(None))
        .distinct()
        .order_by(CS2Item.exterior)
    )
    currencies = session.scalars(
        select(MarketListing.currency_original)
        .where(*base)
        .distinct()
        .order_by(MarketListing.currency_original)
    )
    return ScannerFacets(
        markets=[cast(Platform, value) for value in markets],
        weapons=[value for value in weapons if value is not None],
        exteriors=[value for value in exteriors if value is not None],
        currencies=list(currencies),
    )
